Improve on the error message of Truth.containsExactlyElementsIn

elharo opened the pull request:
fix #970 @netdpb @suztomo

netdpb commented on dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java:
```diff
@@ -74,8 +77,25 @@ public void testReadBom_path() throws MavenRepositoryException, ArtifactDescript
-    if (currentArtifacts.size() != 217) {
-      Truth.assertThat(currentArtifacts).containsExactlyElementsIn(oldArtifacts);
+    if (currentArtifacts.size() != 217) {      
+      Set<Artifact> current = Sets.newHashSet(currentArtifacts);
```
This is a lot of code for something so simple. What about using [Expect](https://truth.dev/api/1.0/com/google/common/truth/Expect.html)?

```java
@Rule public final Expect expect = Expect.create();

...

Set<Artifact> current = Sets.newHashSet(currentArtifacts);
Set<Artifact> old = Sets.newHashSet(oldArtifacts);
expect.withMessage("added artifacts").that(Sets.difference(current, old)).isEmpty();
expect.WithMessage("removed artifacts").that(Sets.difference(old, current)).isEmpty();
```

elharo replied:
I don't think that will show us which artifacts were added and removed though.

netdpb replied:
It will! `assertThat(foo).isEmpty()` shows the unexpected elements on failure. That's part of the design principle of Truth: it tries to show as much useful information as possible.

`assertTrue(foo.isEmpty())` can't do that.

elharo replied:
will it show both or does it fail on the first?

netdpb replied:
It will show both. (Somehow my response appeared above your question.)

elharo commented:
With expect:

1 expectation failed:
  1. added artifacts 
     expected to be empty
     but was: [com.google.cloud:google-cloud-websecurityscanner, com.google.cloud:google-cloud-iot]

     
Without expect:

java.lang.AssertionError: Dependency tree changed. New size is 221
  Added com.google.cloud:google-cloud-websecurityscanner, com.google.cloud:google-cloud-iot

The second seems a little clearer.

netdpb commented on dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java:
```diff
@@ -80,9 +81,13 @@ public void testReadBom_path() throws MavenRepositoryException, ArtifactDescript
       Set<Artifact> old = Sets.newHashSet(oldArtifacts);
-      SetView<Artifact> added = Sets.difference(current, old);
+      
+      Set<String> currentKeys = current.stream().map(Artifacts::makeKey).collect(Collectors.toSet());
```
Why not get the set of strings in one step?

```java
Set<String> current = currentArtifacts.stream().map(Artifacts::makeKey).collect(toSet());
```

elharo replied:
done

netdpb commented on dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java:
```diff
@@ -80,9 +81,13 @@ public void testReadBom_path() throws MavenRepositoryException, ArtifactDescript
       Set<Artifact> old = Sets.newHashSet(oldArtifacts);
-      SetView<Artifact> added = Sets.difference(current, old);
+      
+      Set<String> currentKeys = current.stream().map(Artifacts::makeKey).collect(Collectors.toSet());
```
Don't you want the versions as well? `makeKey()` returns only the `group:artifact`. What about `Artifacts::toCoordinates()`?

elharo replied:
We don't want the versions, because version updates are expected, and typically number in the dozens so the diff is hard to read. What we care about are truly new or removed dependencies.

netdpb commented:
> With expect:
> 
> 1 expectation failed:
> 
> 1. added artifacts
>    expected to be empty
>    but was: [com.google.cloud:google-cloud-websecurityscanner, com.google.cloud:google-cloud-iot]
> 
> Without expect:
> 
> java.lang.AssertionError: Dependency tree changed. New size is 221
> Added com.google.cloud:google-cloud-websecurityscanner, com.google.cloud:google-cloud-iot
> 
> The second seems a little clearer.

Up to you whether that is clearer, or whether the additional clarity is worth adding code that might itself have bugs to your test.

netdpb commented on dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java:
```diff
@@ -101,7 +97,7 @@ public void testReadBom_path() throws MavenRepositoryException, ArtifactDescript
         message += removedMessage;
```
I think a `StringBuilder` or `Formatter` would help.

```java
Formatter message = new Formatter();
formatter.format("Dependency tree changed. New size is %d", currentArtifacts.size());
if (!added.isEmpty()) {
  formatter.format("\n  Added %s", Joiner.on(", ").join(added));
}
if (!removed.isEmpty()) {
  formatter.format("\n  Removed %s", Joiner.on(", ").join(removed));
}
fail(formatter.toString());
```

netdpb commented on dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java:
```diff
@@ -101,7 +97,7 @@ public void testReadBom_path() throws MavenRepositoryException, ArtifactDescript
       }
       
-      if (!added.isEmpty() || ! removed.isEmpty()) {
+      if (!added.isEmpty() || !removed.isEmpty()) {
```
Doesn't at least one of these have to be true, since the current size is not 219?

elharo replied:
makes sense, done
